<a href="https://colab.research.google.com/github/andrewdge/CSE354-Final-Project/blob/main/CSE354_Final_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Document Level Predictions Using Paragraph Level Sentimient

Using the PerSent dataset, we will be training our model based on paragraph-level sentiments. These will be aggregated to produce document-level sentiments.

In [1]:
!pip install transformers

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
import torch
import math
import pandas as pd
from sklearn.model_selection import train_test_split
from torch.utils.data import Dataset, TensorDataset, DataLoader, random_split
from torch.nn.utils.rnn import pad_sequence
from tqdm import tqdm
import numpy as np
import os
from sklearn.metrics import precision_score, recall_score, f1_score
from transformers import DistilBertModel, DistilBertConfig, DistilBertTokenizer, AdamW, DistilBertForSequenceClassification
torch.manual_seed(42)
np.random.seed(42)

# Constants

Constants we will use in our experiments. These may be subjected to change as hyperparameters

In [4]:
DISTILBERT_DROPOUT = 0.2
DISTILBERT_ATT_DROPOUT = 0.2
BATCH_SIZE = 16
EPOCHS = 3

# Andrew PATH
TEST_PATH = '/content/drive/MyDrive/CSE354/random_test.csv'
TRAIN_PATH = '/content/drive/MyDrive/CSE354/train.csv'
VAL_PATH = '/content/drive/MyDrive/CSE354/dev.csv'
SAVE_PATH = '/content/drive/MyDrive/CSE354/models/DistilBERT'

test_data = pd.read_csv(TEST_PATH)
train_data = pd.read_csv(TRAIN_PATH)
# fixed_test will be used for validation
val_data = pd.read_csv(VAL_PATH)



# Initializing Our Model

Here is where we set up our DistilBERT model.

In [5]:
class DistillBERT():
  def __init__(self):
    # TODO(students): start
    self.tokenizer = DistilBertTokenizer.from_pretrained("distilbert-base-uncased")
    config = DistilBertConfig(dropout=DISTILBERT_DROPOUT, 
                          attention_dropout=DISTILBERT_ATT_DROPOUT, 
                          output_hidden_states=True)
    self.model = DistilBertForSequenceClassification.from_pretrained("distilbert-base-uncased", num_labels=3)
    # TODO(students): end
  def get_tokenizer_and_model(self):
    return self.model, self.tokenizer

# DataLoader

This class handles loading, preprocessing, and tokenizing the data.

Each row in the dataframe contains text with some number of paragraphs, as well as a number as labels per paragraph. We add another column in the dataframe, paragraphs per document. This will be used later to test our predictions as compare paragraph-level predictions to paragraph labels, as well as document-level predictions to document labels. We also remove data without paragraph-level labels.

For the labels create new columns for each.

This format is largely takes inspiration from Assignment 3.

In [6]:
class DatasetLoader(Dataset):
  def __init__(self, data, tokenizer):
    # Data is the uncleaned data, as a dataframe.
    self.data = data
    self.tokenizer = tokenizer

  def preprocess_data(self):
    # Combine labels into list.
    df = self.data
    df = df[df['Paragraph0'].notna()]
    df['Paragraph Labels'] = df.iloc[:, 6:].values.tolist() #Includes Nans, remove them
    self.data = df

  def tokenize_data(self):
    # Tokenizing
    tokens = []
    labels = []
    label_dict = {'Negative': 0,
                  'Neutral': 1,
                  'Positive': 2}
    document_list = self.data['DOCUMENT']
    label_list = self.data['Paragraph Labels']

    for (document, doc_labels) in tqdm(zip(document_list, label_list), total=len(document_list)):
      paragraphs = document.split('\n')
      for paragraph, label in zip(paragraphs, doc_labels):
        encoding = self.tokenizer(text=paragraph, truncation='longest_first', max_length=512, return_tensors='pt')
        labels.append(label_dict[label])
        tokens.append(encoding.input_ids[0]) # Might need to CUDA
    
    tokens = pad_sequence(tokens, batch_first=True)
    labels = torch.tensor(labels)
    labels.to("cuda:0" if torch.cuda.is_available() else "cpu")
    tokens.to("cuda:0" if torch.cuda.is_available() else "cpu")
    dataset = TensorDataset(tokens, labels)
    return dataset

  def get_data_loaders(self, shuffle=True):
    self.preprocess_data()
    processed_dataset = self.tokenize_data()
    data_loader = DataLoader(
        processed_dataset,
        shuffle=shuffle,
        batch_size=BATCH_SIZE
    )
    return data_loader

In [7]:
class Trainer():

  def __init__(self, args):
    self.train_data = args['train_data']
    self.val_data = args['val_data']
    self.batch_size = args['batch_size']
    self.epochs = args['epochs']
    self.save_path = args['save_path']
    self.device = args['device']
    transformer = DistillBERT()
    self.model, self.tokenizer = transformer.get_tokenizer_and_model()
    self.model.to(self.device)

  def get_performance_metrics(self, preds, labels):
    pred_flat = np.argmax(preds, axis=1).flatten()
    labels_flat = labels.flatten()
    precision = precision_score(labels_flat, pred_flat, zero_division=0, average='macro')
    recall = recall_score(labels_flat, pred_flat, zero_division=0, average='macro')
    f1 = f1_score(labels_flat, pred_flat, zero_division=0, average='macro')
    return precision, recall, f1

  def train(self, data_loader, optimizer):
    self.model.train()
    total_recall = 0
    total_precision = 0
    total_f1 = 0
    total_loss = 0

    for batch_idx, (reviews, labels) in enumerate(tqdm(data_loader)):
      self.model.zero_grad()
      # TODO(students): start
      output = self.model(reviews.to(self.device), labels=labels.to(self.device)) 
      loss = output.loss
      logits = output.logits
      with torch.no_grad():
        precision, recall, f1 = self.get_performance_metrics(logits.cpu(), labels.cpu())
      loss.backward()
      optimizer.step()
      total_loss += loss
      total_recall += recall
      total_precision += precision
      total_f1 += f1
      if batch_idx % 100 == 0:
        print()
      # TODO(students): end

    precision = total_precision/len(data_loader)
    recall = total_recall/len(data_loader)
    f1 = total_f1/len(data_loader)
    loss = total_loss/len(data_loader)

    return precision, recall, f1, loss

  def eval(self, data_loader):
    self.model.eval()
    total_recall = 0
    total_precision = 0
    total_f1 = 0
    total_loss = 0

    with torch.no_grad():
      for (reviews, labels) in tqdm(data_loader):
        # TODO(students): start
        output = self.model(reviews.to(self.device), labels=labels.to(self.device)) 
        prec, rec, f1 = self.get_performance_metrics(output.logits.cpu(), labels.cpu())
        total_recall += rec
        total_precision += prec
        total_f1 += f1
        total_loss += output.loss
        # TODO(students): end
    
    precision = total_precision/len(data_loader)
    recall = total_recall/len(data_loader)
    f1 = total_f1/len(data_loader)
    loss = total_loss/len(data_loader)

    return precision, recall, f1, loss

  def save_transformer(self):
    self.model.save_pretrained(self.save_path)
    self.tokenizer.save_pretrained(self.save_path)

  def execute(self):
    last_best = 0
    train_dataset = DatasetLoader(self.train_data, self.tokenizer)
    train_data_loader = train_dataset.get_data_loaders(self.batch_size)
    val_dataset = DatasetLoader(self.val_data, self.tokenizer)
    val_data_loader = val_dataset.get_data_loaders(self.batch_size)
    optimizer = AdamW(self.model.parameters(), lr = 3e-5, eps = 1e-8)
    for epoch_i in range(0, self.epochs):
      train_precision, train_recall, train_f1, train_loss = self.train(train_data_loader, optimizer)
      print(f'Epoch {epoch_i + 1}: train_loss: {train_loss:.4f} train_precision: {train_precision:.4f} train_recall: {train_recall:.4f} train_f1: {train_f1:.4f}')
      val_precision, val_recall, val_f1, val_loss = self.eval(val_data_loader)
      print(f'Epoch {epoch_i + 1}: val_loss: {val_loss:.4f} val_precision: {val_precision:.4f} val_recall: {val_recall:.4f} val_f1: {val_f1:.4f}')

      if val_f1 > last_best:
        print("Saving model..")
        # self.save_transformer()
        last_best = val_f1
        print("Model saved.")

In [8]:
import os
import gc
gc.collect()
torch.cuda.empty_cache()
os.environ['CUDA_LAUNCH_BLOCKING'] = "1"
args = {}
args['batch_size'] = BATCH_SIZE
args['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
args['train_data'] = train_data
args['val_data'] = val_data
args['save_path'] = SAVE_PATH
args['epochs'] = EPOCHS
print(args['device'])
CUDA_LAUNCH_BLOCKING=1
trainer = Trainer(args)

trainer.execute()

cuda:0


Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForSequenceClassification: ['vocab_projector.bias', 'vocab_transform.bias', 'vocab_projector.weight', 'vocab_layer_norm.bias', 'vocab_transform.weight', 'vocab_layer_norm.weight']
- This IS expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['pre_classifier.weight', 'classifier.weight', 'pre_clas

  7%|▋         | 101/1498 [01:20<18:23,  1.27it/s]

 13%|█▎        | 201/1498 [02:39<17:06,  1.26it/s]

 20%|██        | 301/1498 [03:58<15:47,  1.26it/s]

 27%|██▋       | 401/1498 [05:17<14:26,  1.27it/s]

 33%|███▎      | 501/1498 [06:36<13:08,  1.26it/s]

 40%|████      | 601/1498 [07:55<11:49,  1.26it/s]

 47%|████▋     | 701/1498 [09:14<10:27,  1.27it/s]

 53%|█████▎    | 801/1498 [10:33<09:11,  1.26it/s]

 60%|██████    | 901/1498 [11:52<07:52,  1.26it/s]

 67%|██████▋   | 1001/1498 [13:11<06:32,  1.27it/s]

 73%|███████▎  | 1101/1498 [14:30<05:13,  1.27it/s]

 80%|████████  | 1201/1498 [15:49<03:54,  1.26it/s]

 87%|████████▋ | 1301/1498 [17:08<02:34,  1.27it/s]

 94%|█████████▎| 1401/1498 [18:27<01:16,  1.27it/s]

100%|██████████| 1498/1498 [19:43<00:00,  1.27it/s]


Epoch 1: train_loss: 0.9542 train_precision: 0.2652 train_recall: 0.3632 train_f1: 0.2747


100%|██████████| 261/261 [01:10<00:00,  3.71it/s]


Epoch 1: val_loss: 0.9374 val_precision: 0.1796 val_recall: 0.3716 val_f1: 0.2384
Saving model..
Model saved.


  0%|          | 1/1498 [00:00<19:35,  1.27it/s]

  7%|▋         | 101/1498 [01:19<18:26,  1.26it/s]

 13%|█▎        | 201/1498 [02:38<17:03,  1.27it/s]

 20%|██        | 301/1498 [03:57<15:47,  1.26it/s]

 27%|██▋       | 401/1498 [05:17<14:27,  1.26it/s]

 33%|███▎      | 501/1498 [06:36<13:09,  1.26it/s]

 40%|████      | 601/1498 [07:55<11:50,  1.26it/s]

 47%|████▋     | 701/1498 [09:14<10:29,  1.27it/s]

 53%|█████▎    | 801/1498 [10:33<09:12,  1.26it/s]

 60%|██████    | 901/1498 [11:52<07:52,  1.26it/s]

 67%|██████▋   | 1001/1498 [13:11<06:32,  1.27it/s]

 73%|███████▎  | 1101/1498 [14:30<05:14,  1.26it/s]

 80%|████████  | 1201/1498 [15:49<03:55,  1.26it/s]

 87%|████████▋ | 1301/1498 [17:09<02:35,  1.26it/s]

 94%|█████████▎| 1401/1498 [18:28<01:16,  1.26it/s]

100%|██████████| 1498/1498 [19:44<00:00,  1.26it/s]


Epoch 2: train_loss: 0.9521 train_precision: 0.2819 train_recall: 0.3638 train_f1: 0.2817


100%|██████████| 261/261 [01:10<00:00,  3.71it/s]


Epoch 2: val_loss: 0.9388 val_precision: 0.1583 val_recall: 0.3646 val_f1: 0.2169


  0%|          | 1/1498 [00:00<19:36,  1.27it/s]

  7%|▋         | 101/1498 [01:19<18:22,  1.27it/s]

 13%|█▎        | 201/1498 [02:38<17:07,  1.26it/s]

 20%|██        | 301/1498 [03:58<15:47,  1.26it/s]

 27%|██▋       | 401/1498 [05:17<14:28,  1.26it/s]

 33%|███▎      | 501/1498 [06:36<13:10,  1.26it/s]

 40%|████      | 601/1498 [07:55<11:48,  1.27it/s]

 47%|████▋     | 701/1498 [09:14<10:27,  1.27it/s]

 53%|█████▎    | 801/1498 [10:33<09:11,  1.26it/s]

 60%|██████    | 901/1498 [11:52<07:53,  1.26it/s]

 67%|██████▋   | 1001/1498 [13:11<06:33,  1.26it/s]

 73%|███████▎  | 1101/1498 [14:31<05:13,  1.27it/s]

 80%|████████  | 1201/1498 [15:50<03:54,  1.27it/s]

 87%|████████▋ | 1301/1498 [17:09<02:35,  1.27it/s]

 94%|█████████▎| 1401/1498 [18:28<01:16,  1.27it/s]

100%|██████████| 1498/1498 [19:44<00:00,  1.26it/s]


Epoch 3: train_loss: 0.9514 train_precision: 0.2719 train_recall: 0.3635 train_f1: 0.2743


100%|██████████| 261/261 [01:09<00:00,  3.77it/s]

Epoch 3: val_loss: 0.9390 val_precision: 0.1770 val_recall: 0.3678 val_f1: 0.2352


Training all layer results:

Epoch 3: train_loss: 0.9514 train_precision: 0.2719 train_recall: 0.3635 train_f1: 0.2743
100%|██████████| 261/261 [01:09<00:00,  3.77it/s]Epoch 3: val_loss: 0.9390 val_precision: 0.1770 val_recall: 0.3678 val_f1: 0.2352